### PR TITLE
document raw-addr-of operators

### PR DESCRIPTION
**Pull request by RalfJung**
as requested by @Havvy 

This is a bit unusual for being exposed as a library feature, fut it really is a fundamental language feature without stable syntax, which is stably accessible through the standard library.

### CODE REVIEW

**Review by Havvy — CHANGES_REQUESTED** [review]
Just woke up and saw this, and it was a great way to start my day. 😍 

I have to go to work, so no time for a full review, but I will at least ask that you switch the style to one line/one sentence to match the rest of the page.

**Comment by RalfJung** [comment]
> I will at least ask that you switch the style to one line/one sentence to match the rest of the page.

Done.

File: src/expressions/operator-expr.md
Comment (Havvy) [inline src/expressions/operator-expr.md:79]:
```suggestion
The raw address-of operators must be used whenever the place expression denotes a place that is not properly aligned or does not store a valid value as determined by its type.
```
```suggestion
The raw address-of operators must be used whenever the place expression could evaluate to a place that is not properly aligned or does not store a valid value as determined by its type.
```

File: src/expressions/operator-expr.md
Comment (Havvy) [inline src/expressions/operator-expr.md:76]: "the expression" here refers to the expression passed to the macro, right? If so, can you name that expression something; perhaps "the referenced operand". Also if so, it needs to be added to the list of place expression contexts on `expressions.md`.

Also, the compare/contrast with borrowing expressions feels weird in the context of the reference. It would make sense in material teaching for the first time, but this isn't a first time reader's document.

File: src/expressions/operator-expr.md
Comment (Havvy) [inline src/expressions/operator-expr.md:82]:
```suggestion
The following is an example of creating a raw pointer to an unaligned place through a `packed` struct:
```

File: src/expressions/operator-expr.md
Comment (Havvy) [inline src/expressions/operator-expr.md:99]:
```suggestion
The following is an example of creating a raw pointer to a place that does not contain a valid value:
```

**Review by Havvy — CHANGES_REQUESTED** [review]

File: src/expressions/operator-expr.md
Comment (RalfJung) [reply src/expressions/operator-expr.md:76]: > "the expression" here refers to the expression passed to the macro, right? If so, can you name that expression something; perhaps "the referenced operand". Also if so, it needs to be added to the list of place expression contexts on expressions.md.

I did some editing here, please check.

> Also, the compare/contrast with borrowing expressions feels weird in the context of the reference. It would make sense in material teaching for the first time, but this isn't a first time reader's document.

I feel like this also makes sense just to increase clarity of the document, no matter how much experience we assume the reader to have. Also note that currently, `### Raw address-of operators` is a *third*-level header, so this is basically "part of" the borrow operator section. The operators are very similar to I felt it would make sense to treat them together, at least to some extend.

File: src/expressions/operator-expr.md
Comment (Havvy) [inline src/expressions/operator-expr.md:76]: The first and third sentences together seem to make up the main definition and then this second sentence is just awkwardly in the middle. Perhaps the second and third sentences should be swapped.

Comment (Havvy) [inline src/expressions/operator-expr.md:76]: It is more clear calling it `expr`, but if you don't name the operand, I will have to when I go through this chapter unifying the style. Your choice if you want to leave it like this or change it.

**Review by Havvy — CHANGES_REQUESTED** [review]
Overall I am happy with the changes. I just see one thing that really stands out to me.

File: src/expressions/operator-expr.md
Comment (RalfJung) [reply src/expressions/operator-expr.md:76]: I don't know what you mean -- the operand, `expr`, is named, isn't it?

Comment (RalfJung) [reply src/expressions/operator-expr.md:76]: I don't know what you mean -- the operand, `expr`, is named, isn't it?

Comment (RalfJung) [reply src/expressions/operator-expr.md]: I made the initial definition self-contained and moved the references to the borrow operator to the second paragraph.

**Comment by RalfJung** [comment]
@Havvy I think I did all the changes you were asking for.

**Comment by RalfJung** [comment]
This PR has been waiting for the next round of review for a while (and the feature it documents is now stable since many releases). How can we make progress?

File: src/expressions/operator-expr.md
Comment (ehuss) [inline src/expressions/operator-expr.md:87]: Does this maybe need to be a well-defined layout like C?  I would assume that the Rust layout could theoretically rearrange the order?

```suggestion
#[repr(C, packed)]
```

File: src/expressions/operator-expr.md
Comment (ehuss) [inline src/expressions/operator-expr.md:75]: It can be helpful to point to the standard library documentation to give more context.

```suggestion
Related to the borrow operators are the *raw address-of operators*, which do not have first-class syntax, but are exposed via the macros [`ptr::addr_of!(expr)`][addr_of] and [`ptr::addr_of_mut!(expr)`][addr_of_mut].
```

And at the bottom:

```
[addr_of]: ../../std/ptr/macro.addr_of.html
[addr_of_mut]: ../../std/ptr/macro.addr_of_mut.html
```

File: src/expressions/operator-expr.md
Comment (ehuss) [inline src/expressions/operator-expr.md:77]: The parenthetical here seems to stick out to me here, and I'm not sure what it is intending to convey.  

```suggestion
`ptr::addr_of!(expr)` then creates a const raw pointer of type `*const T` to the given place, and `ptr::addr_of_mut!(expr)` creates a mutable raw pointer of type `*mut T`.
```

**Comment by ehuss** [comment]
Can you resolve the conflict?

> This PR has been waiting for the next round of review for a while (and the feature it documents is now stable since many releases). How can we make progress?

I don't feel super comfortable hitting the merge button as I am not very familiar with this.  I'd feel better if someone on the lang team approved it.  Perhaps @joshtriplett?

File: src/expressions/operator-expr.md
Comment (RalfJung) [reply src/expressions/operator-expr.md:87]: > I would assume that the Rust layout could theoretically rearrange the order?

It could, but nothing in the example relies on field order. So this is correct without `C`, I think.
We could still add it though, if you want?

Comment (ehuss) [reply src/expressions/operator-expr.md:88]: Ah, makes sense. Leaving as-is seems fine to me.

**Comment by RalfJung** [comment]
@rust-lang/lang it's now been a year since the macros have been stabilized, might be good to also get the reference docs landed at some point. :D

**Comment by joshtriplett** [comment]
I believe all feedback has been addressed at this point. Merging.

**Pull request merged**
